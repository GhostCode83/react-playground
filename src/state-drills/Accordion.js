import React from 'react';

class Accordion extends React.Component {
static defaultProps = {
  sections: []
}

  state = {
    currentSectionIndex: null
  }

  handleButtonClick = (index) => {
    this.setState({ currentSectionIndex: index})
  }

  renderList(section, index, currentSectionIndex) {
    return (
      <li key={index}>
          <button  
            type='submit' 
            onClick={() => this.handleButtonClick(index)}
          >
            {section.title}
          </button>
          {
          (currentSectionIndex === index) && 
          <p>{section.content}</p>}
      </li>)
  }
  
  render() {
    const { currentSectionIndex } = this.state
    const { sections } = this.props
    return (
      <div>
        <ul>
          {sections.map((section, index) =>
          this.renderList(section, index, currentSectionIndex))}
        </ul>
      </div>
    )
  }
}

export default Accordion
