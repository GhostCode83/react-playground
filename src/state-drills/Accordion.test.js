import React from 'react';
import ReactDOM from 'react-dom';
import Accordion from './Accordion';
import renderer from 'react-test-renderer';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

const sectionsProp = [

  {title: "Vegan", 
    content: "plants only"},
  {title: "Octo-Ovo Vegetarian", 
    content: "plants, dairy and eggs"},
  {title: "Omnivour", 
    content: "plants and aninimals"}

]

describe('Accordion Component', () => {
  it('renders without errors', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Accordion />, div);
    ReactDOM.unmountComponentAtNode(div)
  })


  it('renders an empty li when the sections prop is not supplied', () => {
    const tree = renderer
      .create(<Accordion sections={sectionsProp}/>)
      .toJSON()
    expect(tree).toMatchSnapshot();
  })

  it('renders no sections as active by default', () => {
    const wrapper = shallow(<Accordion sections={sectionsProp}/>)
    expect(toJson(wrapper)).toMatchSnapshot();
  })

  it('opens a clicked section', () => {
    const wrapper = shallow(<Accordion sections={sectionsProp} />)
    wrapper.find('button').at(2).simulate('click')
    expect(toJson(wrapper)).toMatchSnapshot()

  })

  it('only opens the last section when multiple sections have been clicked', () => {
    const wrapper = shallow(<Accordion sections={sectionsProp} />)
    wrapper.find('button').at(1).simulate('click')
    wrapper.find('button').at(2).simulate('click')
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})

