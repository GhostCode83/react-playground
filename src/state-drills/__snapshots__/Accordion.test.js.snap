// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Component The component renders an empty li when the sections prop is not supplied 1`] = `
<div>
  <ul>
    <li>
      <button
        onClick={[Function]}
        type="submit"
      >
        Vegan
      </button>
    </li>
    <li>
      <button
        onClick={[Function]}
        type="submit"
      >
        Octo-Ovo Vegetarian
      </button>
    </li>
    <li>
      <button
        onClick={[Function]}
        type="submit"
      >
        Omnivour
      </button>
    </li>
  </ul>
</div>
`;

exports[`Accordion Component only opens the last section when multiple sections have been clicked 1`] = `
<div>
  <ul>
    <li
      key="0"
    >
      <button
        onClick={[Function]}
        type="submit"
      >
        Vegan
      </button>
    </li>
    <li
      key="1"
    >
      <button
        onClick={[Function]}
        type="submit"
      >
        Octo-Ovo Vegetarian
      </button>
    </li>
    <li
      key="2"
    >
      <button
        onClick={[Function]}
        type="submit"
      >
        Omnivour
      </button>
      <p>
        plants and aninimals
      </p>
    </li>
  </ul>
</div>
`;

exports[`Accordion Component opens a clicked section 1`] = `
<div>
  <ul>
    <li
      key="0"
    >
      <button
        onClick={[Function]}
        type="submit"
      >
        Vegan
      </button>
    </li>
    <li
      key="1"
    >
      <button
        onClick={[Function]}
        type="submit"
      >
        Octo-Ovo Vegetarian
      </button>
    </li>
    <li
      key="2"
    >
      <button
        onClick={[Function]}
        type="submit"
      >
        Omnivour
      </button>
      <p>
        plants and aninimals
      </p>
    </li>
  </ul>
</div>
`;

exports[`Accordion Component renders an empty li when the sections prop is not supplied 1`] = `
<div>
  <ul>
    <li>
      <button
        onClick={[Function]}
        type="submit"
      >
        Vegan
      </button>
    </li>
    <li>
      <button
        onClick={[Function]}
        type="submit"
      >
        Octo-Ovo Vegetarian
      </button>
    </li>
    <li>
      <button
        onClick={[Function]}
        type="submit"
      >
        Omnivour
      </button>
    </li>
  </ul>
</div>
`;

exports[`Accordion Component renders no sections as active by default 1`] = `
<div>
  <ul>
    <li
      key="0"
    >
      <button
        onClick={[Function]}
        type="submit"
      >
        Vegan
      </button>
    </li>
    <li
      key="1"
    >
      <button
        onClick={[Function]}
        type="submit"
      >
        Octo-Ovo Vegetarian
      </button>
    </li>
    <li
      key="2"
    >
      <button
        onClick={[Function]}
        type="submit"
      >
        Omnivour
      </button>
    </li>
  </ul>
</div>
`;
